internal/controllers/core/extensionrepo/reconciler.go

```diff
@@ -40,9 +65,116 @@ func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
+		if !ok {
+			return ctrl.Result{}, nil
+		}
+		err := os.RemoveAll(path)
+		if err != nil && !os.IsNotExist(err) {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// Check that the URL is valid.
+	importPath, err := getImportPath(&repo)
+	if err != nil {
+		logger.Get(ctx).Errorf("Invalid ExtensionRepo %s: %v", repo.Name, err)
+		return r.updateStatus(ctx, &repo, func(status *v1alpha1.ExtensionRepoStatus) {
+			status.Error = err.Error()
+		})
+	}
+
+	dlr := get.NewDownloader(r.dlrPath)
+	destPath := dlr.DestinationPath(importPath)
+	info, err := os.Stat(destPath)
+	if err != nil && !os.IsNotExist(err) {
+		return ctrl.Result{}, err
+	}
+
+	// If the destination path does not exist, download it now.
+	if os.IsNotExist(err) {
+		lastFetch := r.fetches[nn]
+		lastBackoff := r.backoffs[nn]
+		if time.Since(lastFetch) < lastBackoff {
+			return ctrl.Result{RequeueAfter: lastBackoff}, nil
+		}
+
+		_, err := dlr.Download(importPath)
+		if err != nil {
+			backoff := r.nextBackoff(nn)
+			backoffMsg := fmt.Sprintf("Downloading ExtensionRepo %s. Waiting %s before retrying. Error: %v", repo.Name, backoff, err)
+			logger.Get(ctx).Errorf("%s", backoffMsg)
+			_, updateErr := r.updateStatus(ctx, &repo, func(status *v1alpha1.ExtensionRepoStatus) {
+				status.Error = backoffMsg
+			})
+			if updateErr != nil {
+				return ctrl.Result{}, updateErr
+			}
+
+			return ctrl.Result{RequeueAfter: backoff}, nil
+		}
+
+		info, err = os.Stat(destPath)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	// Update the status.
+	delete(r.backoffs, nn)
+	r.lastKnownDestPaths[nn] = destPath
+
+	timeFetched := apis.NewTime(info.ModTime())
+	return r.updateStatus(ctx, &repo, func(status *v1alpha1.ExtensionRepoStatus) {
+		status.Error = ""
+		status.LastFetchedAt = timeFetched
+		status.Path = destPath
+	})
+}
+
+// Step up the backoff after an error.
+func (r *Reconciler) nextBackoff(nn types.NamespacedName) time.Duration {
+	backoff := r.backoffs[nn]
+	if backoff == 0 {
+		backoff = 5 * time.Second
+	} else {
+		backoff = 2 * backoff
+	}
+	r.fetches[nn] = time.Now()
+	r.backoffs[nn] = backoff
+	return backoff
+}
+
+// Loosely inspired by controllerutil's Update status algorithm.
+func (r *Reconciler) updateStatus(ctx context.Context, repo *v1alpha1.ExtensionRepo, mutateFn func(*v1alpha1.ExtensionRepoStatus)) (ctrl.Result, error) {
+	update := repo.DeepCopy()
+	mutateFn(&(update.Status))
+
+	if apicmp.DeepEqual(update.Status, repo.Status) {
 		return ctrl.Result{}, nil
 	}
 
-	return ctrl.Result{}, nil
+	err := r.ctrlClient.Status().Update(ctx, update)
+	return ctrl.Result{}, err
+}
+
+func getImportPath(repo *v1alpha1.ExtensionRepo) (string, error) {
+	url := repo.Spec.URL
+	if strings.HasPrefix(url, "https://") {
+		return strings.TrimPrefix(url, "https://"), nil
+	}
+	if strings.HasPrefix(url, "http://") {
+		return strings.TrimPrefix(url, "http://"), nil
+	}
+	return "", fmt.Errorf("malformed url: %v", url)
```

Comment:
Maybe something like "url must start with http:// or https://"? I think it's relatively common for people to (technically incorrectly, but often practically) refer to strings without protocols as urls and the problem might not be obvious from the error message.